test/expr/unary/keypath/keypath.swift

```diff
@@ -893,6 +893,23 @@ struct SR_12290 {
   }
 }
 
+// SR-12745
+struct SR12745 {
+  var x : Int = 0
+}
+
+func testKeyPathHole(_ s: SR12745) {
+  s[\.x] = 10 // expected-error {{cannot infer contextual type for key path; consider adding an explicit root type}}
+  // expected-error@-1 {{value of type 'SR12745' has no subscripts; did you mean to use a key path subscript?}} {{5-5=keyPath: }}
```

Comment:
Thanks for the PR! IIUC, there are three possibilities:

1. The `\.x` does not match any of the explicitly defined subscripts (possibly 0): In this case, the user most likely meant to write an explicit `keyPath: `.
2. The `\.x` matches multiple subscripts which means that the type is ambiguous from the context. It is still possible the the user meant to write an explicit `keyPath:` but less likely than in case 1.
3. There's something else going wrong which is hard to judge (e.g. the user meant to type `x` referring to a local variable instead of `\.x`).

My guess is that case 1 is probably the most common in practice. By adding an extra "cannot infer contextual type for key path; consider adding an explicit root type", that makes it seem like there are 2 different problems to be fixed (the "root type problem" and the "key path subscript problem"), even though just fixing the second one will most likely fix the issue. I think we should only issue the warning about the contextual type here if we are confident that adding the `keyPath:` will not fix the issue.

Reply:
Humm, the thought here was to show all that's going on in the context, because indeed the solver could not infer the root from the context in this case. I think is fine to show it as an error because it is indeed not applicable

Reply:
Ok, now I see where you're coming from. Here's my perspective: I think when a programmer sees an error, they most likely do not want to see all the information available. They want to understand what the error is, why there is an error (i.e. understand the contextual information that is contributing to the error), and hopefully some accurate hints on how to fix it.

In this particular case there's no "one true error", there's multiple possible errors, which is what is making things difficult. If there are multiple possible errors, it should be made clear that we are talking about `error1 || error2`, not `error1 && error2`. IMO showing all of them, unless absolutely necessary, is more likely to lead to frustration (because it makes the programmer think `&&` instead of `||`). Worse, when they see two seemingly unrelated errors at the same spot, and fixing one seemingly fixes both of them, that erodes trust in the compiler. The more often it happens, the more likely the programmer is to start ignoring the diagnostics.

In this particular case, I think there's three main possibilities which I pointed out earlier (maybe there's more which I missed). We can't really address the third one.

Since there is a likelihood that one of the possible errors (first case) is much more likely than the other, I think it's valuable to optimize the diagnostic for that case. This means pruning out inapplicable information from the context (the error for the second case); more information is more room for distraction; if we know that the information is not going to be applicable, we should avoid emitting it.

That said, I don't know how hard implementing the check (to only omit one error instead of both, when applicable) is. I will defer to you and others on that.